sound/soc/sof/pcm.c

```diff
@@ -201,6 +201,14 @@ static int sof_pcm_hw_free(struct snd_pcm_substream *substream)
 	/* send IPC to the DSP */
 	ret = sof_ipc_tx_message(sdev->ipc, stream.hdr.cmd, &stream,
 				 sizeof(stream), &reply, sizeof(reply));
+	/* force power down if FW is in error state */
+	if (ret == -ETIMEDOUT) {
```

Comment:
same question, why just a time-out? have we checked what types of errors we can get here?

Reply:
@plbossart the ETIMEOUT error makes sense because if there are other errors it surely means the DSP is still alive. Its only when the DSP crashes that we run into IPC timeout

Reply:
if the DSP is still alive but throw errors due to a bad programming or configurations, they should not come in the form of a -EXYZ form, they should be a message that's says NAK or something. We are blurring layers if we mix low-level communication issues and protocol issues.

Reply:
@plbossart it actially does send a NAK in the form of setting the ipc reply's error field. The problem is that we are translating it into a -EINVAL.

Reply:
no let me correct myself, -EINVAL is only when the reply size doesnt match the expected reply size. 
Otherwise, it is the error value that is propagated from the DSP. which will certainly be something other than -ETIMEOUT

Reply:
@ranj063 @plbossart My rationale for limiting to ETIMEDOUT is that hw_free can be called in various situations and we could even have invalid call sequences from the ALSA core (or introduced by user somehow) and power recycling on these case might lead to hard to debug scenarios.

ETIMEDOUT OTOH seems a safe bet. The FW is alive and it should always respond to the IPC. If it is not, it is definitely not working as specific and semantically at 'hw_free', power cycling the FW seems safe.

Maybe this is too conservative, but it is safe and handles a large number of real-life scenarios.